File: samples/python/speech_generation/kokoro_phonemize_fallback.py
#!/usr/bin/env python3
# Copyright (C) 2025 Intel Corporation
# SPDX-License-Identifier: Apache-2.0

import argparse

import numpy as np
import openvino as ov
import openvino_genai
import soundfile as sf


def _load_kokoro_embedding(file_path: str, shape):
    """Load a float32 binary and reshape it according to the given ov.Shape."""
    data = np.fromfile(file_path, dtype=np.float32)
    if data.size == 0:
        raise RuntimeError(f"Speaker embedding file is empty: {file_path}")
    return ov.Tensor(data.reshape(shape))


def main():
    parser = argparse.ArgumentParser()
    parser.add_argument("model_dir", help="Path to Kokoro model directory")
    parser.add_argument("text", help="Input text for speech generation")
    parser.add_argument(
        "--speaker_embedding_file_path", required=True, help="Path to a prepared Kokoro speaker embedding binary"
    )
    parser.add_argument(
        "--language",
        default="en-us",
        choices=["en-us", "en-gb"],
        help="Kokoro language variant for this sample (en-us or en-gb)",
    )
    parser.add_argument("--device", default="CPU", help="Device to run the model on (default: CPU)")
    parser.add_argument(
        "--phonemize_fallback_model_dir",
        default=None,
        help=(
            "Optional OpenVINO fallback G2P model directory. "
            "If set, Kokoro phonemize fallback uses this model. "
            "If omitted, fallback defaults to espeak-ng."
        ),
    )
    parser.add_argument("--output", default="output_audio.wav", help="Output WAV file path")
    args = parser.parse_args()

    pipe = openvino_genai.Text2SpeechPipeline(args.model_dir, args.device)
    speaker_embedding = _load_kokoro_embedding(args.speaker_embedding_file_path, pipe.get_speaker_embedding_shape())

    config = pipe.get_generation_config()
    config.language = args.language
    config.phonemize_fallback_model_dir = args.phonemize_fallback_model_dir
    pipe.set_generation_config(config)

    result = pipe.generate(args.text, speaker_embedding)
    assert len(result.speeches) == 1, "Expected one waveform"

    speech = result.speeches[0]
    speech_data = np.array(speech.data).reshape(-1)
    sf.write(args.output, speech_data, samplerate=result.output_sample_rate)

    fallback_mode = (
        f"OpenVINO fallback ({args.phonemize_fallback_model_dir})"
        if args.phonemize_fallback_model_dir
        else "espeak-ng fallback"
    )
    print(f"[Info] Generated '{args.output}' using {fallback_mode} during phonemize/G2P fallback.")


if __name__ == "__main__":
    main()

File: samples/cpp/speech_generation/kokoro_phonemize_fallback.cpp
// Copyright (C) 2023-2026 Intel Corporation
// SPDX-License-Identifier: Apache-2.0

#include "audio_utils.hpp"
#include "openvino/genai/speech_generation/text2speech_pipeline.hpp"

#include <optional>
#include <string>

int main(int argc, char* argv[]) try {
    OPENVINO_ASSERT(
        argc >= 3,
        "Usage: ",
        argv[0],
        " <MODEL_DIR> \"<PROMPT>\" [--voice <VOICE_ID>] [--language <en-us|en-gb>] "
        "[--phonemize_fallback_model_dir <DIR>] [--output <FILE.wav>] [--device <DEVICE>]");

    const std::string models_path = argv[1];
    const std::string prompt = argv[2];

    std::string device = "CPU";
    std::string voice = "af_heart";
    std::string language = "en-us";
    std::optional<std::string> phonemize_fallback_model_dir;
    std::string output_file_name = "output_audio.wav";

    int idx = 3;
    while (idx < argc) {
        const std::string option = argv[idx++];
        OPENVINO_ASSERT(idx < argc, "Missing value for option ", option);
        const std::string value = argv[idx++];

        if (option == "--voice") {
            voice = value;
        } else if (option == "--language") {
            language = value;
        } else if (option == "--phonemize_fallback_model_dir") {
            phonemize_fallback_model_dir = value;
        } else if (option == "--output") {
            output_file_name = value;
        } else if (option == "--device") {
            device = value;
        } else {
            OPENVINO_THROW("Unknown option: ", option);
        }
    }

    OPENVINO_ASSERT(language == "en-us" || language == "en-gb",
                    "kokoro_phonemize_fallback sample supports only --language en-us or en-gb");

    ov::genai::Text2SpeechPipeline pipe(models_path, device);
    auto config = pipe.get_generation_config();
    config.voice = voice;
    config.language = language;
    config.phonemize_fallback_model_dir = phonemize_fallback_model_dir;
    pipe.set_generation_config(config);

    auto result = pipe.generate(prompt, ov::Tensor());
    OPENVINO_ASSERT(result.speeches.size() == 1, "Expected exactly one decoded waveform");

    const auto waveform_size = result.speeches[0].get_size();
    const auto* waveform_ptr = result.speeches[0].data<const float>();
    const auto bits_per_sample = result.speeches[0].get_element_type().bitwidth();
    utils::audio::save_to_wav(waveform_ptr,
                              waveform_size,
                              output_file_name,
                              bits_per_sample,
                              result.output_sample_rate);

    std::cout << "[Info] Saved: " << output_file_name << std::endl;
    if (phonemize_fallback_model_dir.has_value()) {
        std::cout << "[Info] Phonemize fallback: OpenVINO model at '" << *phonemize_fallback_model_dir << "'"
                  << std::endl;
    } else {
        std::cout << "[Info] Phonemize fallback: espeak-ng" << std::endl;
    }

    return EXIT_SUCCESS;
} catch (const std::exception& error) {
    try {
        std::cerr << error.what() << '\n';
    } catch (const std::ios_base::failure&) {
    }
    return EXIT_FAILURE;
} catch (...) {
    try {
        std::cerr << "Non-exception object thrown\n";
    } catch (const std::ios_base::failure&) {
    }
    return EXIT_FAILURE;
}
